src/utils/useIsomorphicLayoutEffect.js

```diff
@@ -12,6 +13,7 @@ import { useEffect, useLayoutEffect } from 'react'
 export const useIsomorphicLayoutEffect =
   typeof window !== 'undefined' &&
   typeof window.document !== 'undefined' &&
-  typeof window.document.createElement !== 'undefined'
+  typeof window.document.createElement !== 'undefined' &&
+  !Server.isServerSide()
```

Comment:
An other option that would be simpler, rather than importing wonder blocks to support this, would be to detect if we're running in jsdom, or not, following this: https://github.com/jsdom/jsdom/issues/1537 That seems to match the intent of the comment above, as well.

Reply:
That would result in client-side tests also using `useEffect` whereas they'd normally use `useLayoutEffect`.  I'm not sure what impact that would have on the tests, but it seems like it's safer to check `Server.isServerSide()`.

Reply:
Why would that affect the client-side tests? the jsdom check would be false, like the isServerSide check would be

Reply:
@testing-library uses jsdom under the hood so I'd expect the jsdom check to be true for all tests using @testing-library.  I'll check to see if that's actually the case.

Reply:
`navigator.userAgent` for a client-side test is `Mozilla/5.0 (darwin) AppleWebKit/537.36 (KHTML, like Gecko) jsdom/16.7.0`.

Reply:
Presumably we could also look at the NODE_ENV to determine that we're in a testing environment? But I guess at that point we're going to have issues with actually testing this code? This just feels *really* bad, we really don't want to be calling our custom code from inside internal libraries - we went down this road in the past and it made it really hard for us to disentangle.

Reply:
An alternative approach might be for us to add an `ssr` prop to `Provider` and have it pass that value to the `connect()` and the various hooks and they can pass it to `useIsomorphicLayoutEffect()` calls.  It's more work, but feels like something that could possible by upstreamed at some point.

Reply:
That sounds interesting... what would the work be involved with that? I agree that it sounds like something that could be upstreamed and would make it much easier for us to maintain.

Reply:
I don't think it should be too bad.  There are only three files that make use of `useIsomorphicLayoutEffect`:
- connectAdvanced.js
- Provider.js
- useSelector.js

Adding `ssr` to the Context object that Provider.js provides to everything shouldn't be difficult.  I think we'll want to rework `useIsomorphicLayoutEffect` to instead be `getUseIsomorphicLayoutEffect = (ssr: boolean) => ...`, but that shouldn't add too much friction.

Reply:
Our use of `Provider` will then become:
```
<Provider store={store} ssr={Server.isServerSide()}>
    ...
</Provider>
```

Reply:
I starting working on this and it turns out this is a bit trickier than I thought since using the `ssr` prop requires using an `if/else` block to determine which hook to run which is a no-no as far as the rules of hooks is concerned so now I need to create a hook that wraps the decision logic.

Reply:
I don't think there's a way for me to wrap that logic in a hook.  Instead I'm going to provide a setter that allows people to override the default logic when deciding whether we're SSR-ing or not.  Here's a sketch of it:
```
const defaultShouldSSR = () => (
    typeof window !== 'undefined' &&
    typeof window.document !== 'undefined' &&
    typeof window.document.createElement !== 'undefined'
);

let _shouldSSR = defaultShouldSSR;

export const shouldSSR = () => _shouldSSR();

export const setShouldSSR = (fn) => _shouldSSR = fn;
export const resetShouldSSR = () => _shouldSSR = defaultShouldSSR;
```
We'd export `setShouldSSR` at the top-level and then using it like so in the route tests:
```
beforeEach(() => {
   setShouldSSR(() => Server.isServerSide());
}

afterEach(() => {
  resetShouldSSR();
}

...
```

Reply:
Ok yeah, that seems a bit more reasonable! It makes the dependency less strict, at least - much easier to maintain!

Reply:
That's not going to work either because `setShouldSSR()` is going to get called after the Provider.js _et al_ import useIsomorphicLayoutEffect.js.  I need to do some more thinking on this.

Reply:
Maybe it will work, but we'll still need to change `useIsomorphicLayoutEffect` to return a getter function.  It should be "safe" from a "rule of hooks" perspective since we'll never be changing what `shouldSSR()` while the application (or component-under-test) is mounted.  I'll experiment with that.